### PR TITLE
Remove/comment new uses of `not_expecting` in the parser

### PR DESCRIPTION
**Pull request by ncik-roberts**
**Review:** @ccasin (as discussed); this is low-urgency

Using the function `not_expecting` to raise syntax errors is discouraged in a comment in the parser. It makes some menhir analyses misbehave and it interacts poorly with tools. (E.g. merlin defines `not_expecting` as a unit-returning function, so the merlin code needs to produce an AST in these cases, which complicates the merge.)

I don't think any of these are huge problems, but we should probably avoid using it in new code, which means making sure we're commenting any existing use. I took a few approaches in this PR:
  * comment (what I hope to be) short-lived use sites, e.g. the special-cased parsing of `float#`
  * refactor grammar to turn ad-hoc `not_expecting` syntax errors into normal menhir syntax errors

### CODE REVIEW

**Review by ccasin — APPROVED** [review]
Looks like a clear improvement - thanks!

**Comment by ccasin** [comment]
(Unfortunately, because I let it sit for a couple weeks, rebasing is needed before it can merge - sorry)

**Comment by ncik-roberts** [comment]
No problem, the fact that I also let it set for 3 weeks doesn't suggest this one was that urgent. I'll merge when CI turns green.

**Pull request merged**
